### PR TITLE
feat: message withdraw emissions

### PR DESCRIPTION
**Pull request by kingpinXD**
# Description
Add a message to enable observers to collect emission rewards
Closes: https://github.com/zeta-chain/node/issues/1653

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Include instructions and any relevant details so others can reproduce. 

- [ ] Tested CCTX in localnet
- [ ] Tested in development environment
- [ ] Go unit tests
- [ ] Go integration tests
- [ ] Tested via GitHub Actions 

# Checklist:

- [ ] I have added unit tests that prove my fix feature works

### CODE REVIEW

**Comment by akimov99900** [comment]
nice code

File: x/emissions/client/cli/tx_create_withdraw_emssions.go [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+	sdkmath "cosmossdk.io/math"
+	"github.com/cosmos/cosmos-sdk/client"
+	"github.com/cosmos/cosmos-sdk/client/flags"
+	"github.com/cosmos/cosmos-sdk/client/tx"
+	"github.com/spf13/cobra"
+	"github.com/zeta-chain/zetacore/x/emissions/types"
+)
+
+func CmdCreateWithdrawEmission() *cobra.Command {
```
Comment (lumtis) [inline x/emissions/client/cli/tx_create_withdraw_emssions.go:14]: Why "create"?

We should name the CLI as the action of the message `WithdrawEmission`

File: x/emissions/types/errors.go [hunk 2]
```diff
@@ -1,14 +1,13 @@
...
-// x/emissions module sentinel errors
 var (
-	ErrEmissionTrackerNotFound = sdkerrors.Register(ModuleName, 1100, "Emission Tracker Not found")
-	ErrParsingSenderAddress    = sdkerrors.Register(ModuleName, 1101, "Unable to parse address of sender")
-	ErrAddingCoinstoTracker    = sdkerrors.Register(ModuleName, 1102, "Unable to add coins to emissionTracker ")
+	ErrEmissionsNotFound                   = errorsmod.Register(ModuleName, 1000, "Emissions not found")
+	ErrUnableToWithdrawEmissions           = errorsmod.Register(ModuleName, 1002, "Unable to withdraw emissions")
+	ErrInvalidAddress                      = errorsmod.Register(ModuleName, 1003, "Invalid address")
+	ErrRewardsPoolDoesNotHaveEnoughBalance = errorsmod.Register(ModuleName, 1004, "Rewards pool does not have enough balance")
+	ErrInvalidAmount                       = errorsmod.Register(ModuleName, 1005, "Invalid amount")
```
Comment (lumtis) [inline x/emissions/types/errors.go:12]: We should use full lower case for errors

File: x/emissions/types/message_withdraw_emissions.go [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+	bz := ModuleCdc.MustMarshalJSON(msg)
+	return sdk.MustSortJSON(bz)
+}
+
+func (msg *MsgWithdrawEmission) ValidateBasic() error {
+	_, err := sdk.AccAddressFromBech32(msg.Creator)
+	if err != nil {
+		return errorsmod.Wrapf(sdkerrors.ErrInvalidAddress, "invalid creator address (%s)", err)
+	}
+	if !msg.Amount.GT(sdkmath.ZeroInt()) {
```
Comment (lumtis) [inline x/emissions/types/message_withdraw_emissions.go:46]: We should also check `IsNil`

File: x/emissions/types/keys.go [hunk 4]
```diff
@@ -23,6 +23,7 @@ const (
 	// MemStoreKey defines the in-memory store key
 	MemStoreKey              = "mem_emissions"
 	WithdrawableEmissionsKey = "WithdrawableEmissions-value-"
+	WithdrawEmissionsKey     = "WithdrawEmissions-value-"
```
Comment (lumtis) [inline x/emissions/types/keys.go:26]:
```suggestion
```

File: x/emissions/types/errors.go [hunk 5]
```diff
@@ -1,14 +1,13 @@
 package types
 
-// DONTCOVER
+import errorsmod "cosmossdk.io/errors"
 
-import (
-	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
-)
+// DONTCOVER
```
Comment (lumtis) [inline x/emissions/types/errors.go:5]:
```suggestion
```
Already added in codecov.yml that we don't cover this file

File: x/emissions/keeper/msg_server_withdraw_emissions.go [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+// All withdraws for a block are deleted at the end of the block irrespective of whether they were processed or not.
+func (k msgServer) WithdrawEmission(goCtx context.Context, msg *types.MsgWithdrawEmission) (*types.MsgWithdrawEmissionResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// check if the creator address is valid
+	address, err := sdk.AccAddressFromBech32(msg.Creator)
+	if err != nil {
+		return nil, errorsmod.Wrap(types.ErrInvalidAddress, err.Error())
+	}
+	// check if the undistributed rewards pool has enough balance to process this request.
```
Comment (lumtis) [inline x/emissions/keeper/msg_server_withdraw_emissions.go:26]:
```suggestion
	}
	
	// check if the undistributed rewards pool has enough balance to process this request.
```

File: x/emissions/keeper/msg_server_withdraw_emissions.go [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+	undistributedRewardsBalance := k.GetBankKeeper().GetBalance(ctx, types.UndistributedObserverRewardsPoolAddress, config.BaseDenom)
+	if undistributedRewardsBalance.Amount.LT(msg.Amount) {
+		return nil, errorsmod.Wrap(types.ErrRewardsPoolDoesNotHaveEnoughBalance, " rewards pool does not have enough balance to process this request")
+	}
+
+	err = k.RemoveWithdrawableEmission(ctx, msg.Creator, msg.Amount)
+	if err != nil {
+		return nil, errorsmod.Wrap(types.ErrUnableToWithdrawEmissions, err.Error())
+	}
+	err = k.GetBankKeeper().SendCoinsFromModuleToAccount(ctx, types.UndistributedObserverRewardsPool, address, sdk.NewCoins(sdk.NewCoin(config.BaseDenom, msg.Amount)))
```
Comment (lumtis) [inline x/emissions/keeper/msg_server_withdraw_emissions.go:37]:
```suggestion
	}
	
	err = k.GetBankKeeper().SendCoinsFromModuleToAccount(ctx, types.UndistributedObserverRewardsPool, address, sdk.NewCoins(sdk.NewCoin(config.BaseDenom, msg.Amount)))
```

File: x/emissions/keeper/msg_server_withdraw_emissions.go [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+	if err != nil {
+		ctx.Logger().Error(fmt.Sprintf("Error while processing withdraw of emission to adresss %s for amount %s : err %s", address, msg.Amount, err))
+		return nil, errorsmod.Wrap(types.ErrUnableToWithdrawEmissions, err.Error())
+	}
+	return &types.MsgWithdrawEmissionResponse{}, nil
```
Comment (lumtis) [inline x/emissions/keeper/msg_server_withdraw_emissions.go:42]:
```suggestion
	}
	
	return &types.MsgWithdrawEmissionResponse{}, nil
```

File: x/emissions/keeper/msg_server_withdraw_emissions.go [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+// WithdrawEmission create a withdraw emission object , which is then process at endblock
+// The withdraw emission object is created and stored
+// using the address of the creator as the index key ,therefore, if more that one withdraw requests are created in a block on thr last one would be processed.
+// Creating a withdraw does not guarantee that the emission will be processed
+// All withdraws for a block are deleted at the end of the block irrespective of whether they were processed or not.
+func (k msgServer) WithdrawEmission(goCtx context.Context, msg *types.MsgWithdrawEmission) (*types.MsgWithdrawEmissionResponse, error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	// check if the creator address is valid
+	address, err := sdk.AccAddressFromBech32(msg.Creator)
```
Comment (lumtis) [inline x/emissions/keeper/msg_server_withdraw_emissions.go:22]: Already checked in the message validation. 

If we consider this state transition could be called outside of external message, we can replace this check by calling `msg.Validate` instead

File: x/emissions/keeper/msg_server_withdraw_emissions.go [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+	// check if the undistributed rewards pool has enough balance to process this request.
+	// This is just a preliminary check, the actual processing at endblock might still fail if the pool balance gets affected.
+	undistributedRewardsBalance := k.GetBankKeeper().GetBalance(ctx, types.UndistributedObserverRewardsPoolAddress, config.BaseDenom)
+	if undistributedRewardsBalance.Amount.LT(msg.Amount) {
+		return nil, errorsmod.Wrap(types.ErrRewardsPoolDoesNotHaveEnoughBalance, " rewards pool does not have enough balance to process this request")
+	}
+
+	err = k.RemoveWithdrawableEmission(ctx, msg.Creator, msg.Amount)
+	if err != nil {
+		return nil, errorsmod.Wrap(types.ErrUnableToWithdrawEmissions, err.Error())
```
Comment (lumtis) [inline x/emissions/keeper/msg_server_withdraw_emissions.go:35]:
```suggestion
		return nil, errorsmod.Wrap(types.ErrRemoveWithdrawableEmission, err.Error())
```

File: x/emissions/keeper/msg_server_withdraw_emissions.go (hunk 9 above)
Comment (kingpinXD) [reply x/emissions/keeper/msg_server_withdraw_emissions.go:22]: I do not see any harm in keeping the check, we need to parse the address anyway

File: x/emissions/client/cli/tx_create_withdraw_emssions.go (hunk 1 above)
Comment (kingpinXD) [reply x/emissions/client/cli/tx_create_withdraw_emssions.go]: f2d4d84768a74f837eece318f4b2c9fa557c3f81

File: x/emissions/types/errors.go (hunk 2 above)
Comment (kingpinXD) [reply x/emissions/types/errors.go]: f2d4d84768a74f837eece318f4b2c9fa557c3f81

File: x/emissions/types/message_withdraw_emissions.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (kingpinXD) [reply x/emissions/types/message_withdraw_emissions.go]: f2d4d84768a74f837eece318f4b2c9fa557c3f81

Comment (kingpinXD) [reply x/emissions/types/message_withdraw_emissions.go]: f2d4d84768a74f837eece318f4b2c9fa557c3f81

File: x/emissions/types/message_withdraw_emissions.go [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+		return errorsmod.Wrapf(ErrInvalidAmount, "invalid amount (%s)", msg.Amount.String())
```
[the remark this answers is not in the document]
Comment (kingpinXD) [reply x/emissions/types/message_withdraw_emissions.go]: f2d4d84768a74f837eece318f4b2c9fa557c3f81

File: x/emissions/types/errors.go (hunk 5 above)
Comment (kingpinXD) [reply x/emissions/types/errors.go]: f2d4d84768a74f837eece318f4b2c9fa557c3f81

File: x/emissions/keeper/msg_server_withdraw_emissions.go (hunk 10 above)
Comment (kingpinXD) [reply x/emissions/keeper/msg_server_withdraw_emissions.go]: f2d4d84768a74f837eece318f4b2c9fa557c3f81

File: testutil/keeper/keeper.go [hunk 12]
```diff
@@ -106,7 +106,7 @@ var moduleAccountPerms = map[string][]string{
 	crosschaintypes.ModuleName:                      {authtypes.Minter, authtypes.Burner},
 	fungibletypes.ModuleName:                        {authtypes.Minter, authtypes.Burner},
 	emissionstypes.ModuleName:                       {authtypes.Minter},
-	emissionstypes.UndistributedObserverRewardsPool: nil,
+	emissionstypes.UndistributedObserverRewardsPool: {authtypes.Minter, authtypes.Burner},
```
[the remark this answers is not in the document]
Comment (kingpinXD) [reply testutil/keeper/keeper.go]: f29adf275015580bcf348542254f38ec4bf5244f

File: x/emissions/keeper/msg_server_withdraw_emissions.go (hunk 9 above)
Comment (lumtis) [reply x/emissions/keeper/msg_server_withdraw_emissions.go:22]: Ok, fine, didn't realize address is used

File: x/emissions/types/keys.go (hunk 4 above)
Comment (lumtis) [reply x/emissions/types/keys.go]: Removing this one as no longer used FYI

**Review by lumtis — APPROVED** [review]

File: x/emissions/types/errors.go [hunk 13]
```diff
@@ -1,14 +1,11 @@
 package types
 
-// DONTCOVER
```
[the remark this answers is not in the document]
Comment (lumtis) [reply x/emissions/types/errors.go:3]: We can remove all of them, it's already ignore with the `codecov.yml` (not for local coverage but I think we can rely on codecov in any case)

File: x/emissions/keeper/withdrawable_emissions.go [hunk 14]
```diff
@@ -46,7 +48,27 @@ func (k Keeper) AddObserverEmission(ctx sdk.Context, address string, amount sdkm
 	k.SetWithdrawableEmission(ctx, we)
 }
 
+// RemoveWithdrawableEmission removes the given amount from the withdrawable emission of a given address.
+// If the amount is greater than the withdrawable emission, it will remove the entire withdrawable emission.
```
[the remark this answers is not in the document]
Comment (kingpinXD) [reply x/emissions/keeper/withdrawable_emissions.go]: Not sure what you mean, I changed it up a little bit , lmk if it sounds better now 
3cc3181c4c5bd628e35f9784424f2cc7a42af2d8

File: x/emissions/keeper/msg_server_withdraw_emissions_test.go [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+		err = sk.BankKeeper.SendCoinsFromModuleToModule(ctx, types.ModuleName, types.UndistributedObserverRewardsPool, sdk.NewCoins(sdk.NewCoin(config.BaseDenom, withdrawableEmission.Amount)))
+		require.NoError(t, err)
+
+		_, err = msgServer.WithdrawEmission(ctx, &types.MsgWithdrawEmission{
+			Creator: withdrawableEmission.Address,
+			Amount:  withdrawableEmission.Amount,
+		})
+		require.NoError(t, err)
+
+		balance := k.GetBankKeeper().GetBalance(ctx, sdk.MustAccAddressFromBech32(withdrawableEmission.Address), config.BaseDenom).Amount.String()
```
[the remark this answers is not in the document]
Comment (kingpinXD) [reply x/emissions/keeper/msg_server_withdraw_emissions_test.go:35]: 3cc3181c4c5bd628e35f9784424f2cc7a42af2d8

File: x/emissions/client/cli/tx_withdraw_emssions.go [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+
+	sdkmath "cosmossdk.io/math"
+	"github.com/cosmos/cosmos-sdk/client"
+	"github.com/cosmos/cosmos-sdk/client/flags"
+	"github.com/cosmos/cosmos-sdk/client/tx"
+	"github.com/spf13/cobra"
+	"github.com/zeta-chain/zetacore/x/emissions/types"
+)
+
+func CmdWithdrawEmission() *cobra.Command {
```
[the remark this answers is not in the document]
Comment (kingpinXD) [reply x/emissions/client/cli/tx_withdraw_emssions.go]: 3cc3181c4c5bd628e35f9784424f2cc7a42af2d8

File: docs/spec/emissions/messages.md [hunk 17]
```diff
@@ -0,0 +1,17 @@
+# Messages
+
+## MsgWithdrawEmission
+
+WithdrawEmission create a withdraw emission object , which is then process at endblock
```
[the remark this answers is not in the document]
Comment (kingpinXD) [reply docs/spec/emissions/messages.md]: 3cc3181c4c5bd628e35f9784424f2cc7a42af2d8
Thanks for checking

File: x/emissions/keeper/msg_server_withdraw_emissions_test.go [hunk 18]
```diff
@@ -0,0 +1,115 @@
...
+		withdrawableEmission := sample.WithdrawableEmissions(t)
+		k.SetWithdrawableEmission(ctx, withdrawableEmission)
+		_, err := msgServer.WithdrawEmission(ctx, &types.MsgWithdrawEmission{
+			Creator: withdrawableEmission.Address,
+			Amount:  sdkmath.NewInt(-1),
+		})
+		require.ErrorIs(t, err, types.ErrUnableToWithdrawEmissions)
+	})
+
+	t.Run("unable to withdraw emissions if SendCoinsFromModuleToAccount", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (kingpinXD) [reply x/emissions/keeper/msg_server_withdraw_emissions_test.go]: 1ae19df1deb8fe0ee9e48d226e0eff09f328cbb0

**Pull request merged**